src/Parser/Druid/Restoration/Modules/Core/HotTracking/HotTracker.js

```diff
@@ -0,0 +1,205 @@
+    const oneStack = this.mastery.decomposeHeal(event).oneStack; // TODO move this before target validation so still counts direct healing
+    Object.keys(this.hots[targetId]).forEach(otherSpellId => {
+      if (otherSpellId !== spellId) {
+        const otherHot = this.hots[targetId][otherSpellId];
+        otherHot.attributions.forEach(att => att.masteryHealing += oneStack);
+        // boosts don't get mastery benefit because the hot was there with or without the boost
+        // TODO add handling for HoT extensions
+      }
+    });
+  }
+
+  on_byPlayer_applybuff(event) {
+    // if (event.ability.guid === SPELLS.AWAKENED.id) {
+    //   // 4t21 proc causes 5 dreamer procs that would not otherwise have happened
+    //   // while in reality there is usually a quick sequence of 6 dreamers, one of which is "natural",
+    //   // we simplify by attributing the next 5 to 4t21
+    //   this.remaining4t21attributes = 5;
+    // }
+
+    const spellId = event.ability.guid;
+    const target = this._validateAndGetTarget(event);
+    if (!target) {
+      return;
+    }
+    const targetId = event.targetID;
+
+    const newHot = {
+      start: event.timestamp,
+      end: event.timestamp + this.hotInfo[spellId].duration,
+      ticks: [], // listing of ticks w/ effective heal amount and timestamp, to be used as part of the HoT extension calculations
+      attributions: [], // The effect or bonus that procced this HoT application. No attribution implies the spell was hardcast.
+      extensions: [], // The effects or bonuses that caused this HoT to have extended duration. TODO NYI
+      boosts: [], // The effects or bonuses that caused the strength of this HoT to be boosted for its full duration.
+    };
+    if(!this.hots[targetId]) {
+      this.hots[targetId] = {};
+    }
+    this.hots[targetId][spellId] = newHot;
+  }
+
+  on_byPlayer_refreshbuff(event) {
+    const spellId = event.ability.guid;
+    const target = this._validateAndGetTarget(event);
+    if (!target) {
+      return;
+    }
+    const targetId = event.targetID;
+
+    const hot = this.hots[targetId][spellId];
+    const oldEnd = hot.end;
+
+    // TODO check / build suggestion for early refreshes, etc
+
+    const newEndMax = oldEnd + this.hotInfo[spellId].duration;
+    const pandemicMax = event.timestamp + (this.hotInfo[spellId].duration * PANDEMIC_FACTOR);
+    hot.end = Math.min(newEndMax, pandemicMax);
+
+    hot.attributions = []; // new attributions on refresh
+    // TODO add handling for HoT extensions... how should extensions be handled in the case of an early refresh?
+    hot.boosts = [];
+  }
+
+  on_byPlayer_removebuff(event) {
+    if(event.ability.guid === SPELLS.POWER_OF_THE_ARCHDRUID_BUFF.id) {
+      this.potaFallTimestamp = event.timestamp; // FIXME temp until figure out hasBuff problem
+    }
+
+    const spellId = event.ability.guid;
+    const target = this._validateAndGetTarget(event);
```

Comment:
why don't you just move ``const spellId = event.ability.guid;`` that initial check so that could be if ``spellID ===`` aswell?